File: software/examples/ruby/example_callback.rb
#!/usr/bin/env ruby
# -*- ruby encoding: utf-8 -*-

require 'tinkerforge/ip_connection'
require 'tinkerforge/bricklet_air_quality'

include Tinkerforge

HOST = 'localhost'
PORT = 4223
UID = 'XYZ' # Change XYZ to the UID of your Air Quality Bricklet

ipcon = IPConnection.new # Create IP connection
aq = BrickletAirQuality.new UID, ipcon # Create device object

ipcon.connect HOST, PORT # Connect to brickd
# Don't use device before ipcon is connected

# Register all values callback
aq.register_callback(BrickletAirQuality::CALLBACK_ALL_VALUES) do |iaq_index,
                                                                  iaq_index_accuracy,
                                                                  temperature, humidity,
                                                                  air_pressure|
  puts "IAQ Index: #{iaq_index}"

  if iaq_index_accuracy == BrickletAirQuality::ACCURACY_UNRELIABLE
    puts "IAQ Index Accuracy: Unreliable"
  elsif iaq_index_accuracy == BrickletAirQuality::ACCURACY_LOW
    puts "IAQ Index Accuracy: Low"
  elsif iaq_index_accuracy == BrickletAirQuality::ACCURACY_MEDIUM
    puts "IAQ Index Accuracy: Medium"
  elsif iaq_index_accuracy == BrickletAirQuality::ACCURACY_HIGH
    puts "IAQ Index Accuracy: High"
  end

  puts "Temperature: #{temperature/100.0} °C"
  puts "Humidity: #{humidity/100.0} %RH"
  puts "Air Pressure: #{air_pressure/100.0} mbar"
  puts ''
end

# Set period for all values callback to 1s (1000ms)
aq.set_all_values_callback_configuration 1000, false

puts 'Press key to exit'
$stdin.gets
ipcon.disconnect

File: software/examples/perl/example_callback.pl
#!/usr/bin/perl

use strict;
use Tinkerforge::IPConnection;
use Tinkerforge::BrickletAirQuality;

use constant HOST => 'localhost';
use constant PORT => 4223;
use constant UID => 'XYZ'; # Change XYZ to the UID of your Air Quality Bricklet

# Callback subroutine for all values callback
sub cb_all_values
{
    my ($iaq_index, $iaq_index_accuracy, $temperature, $humidity, $air_pressure) = @_;

    print "IAQ Index: $iaq_index\n";

    if ($iaq_index_accuracy == Tinkerforge::BrickletAirQuality->ACCURACY_UNRELIABLE)
    {
        print "IAQ Index Accuracy: Unreliable\n";
    }
    elsif ($iaq_index_accuracy == Tinkerforge::BrickletAirQuality->ACCURACY_LOW)
    {
        print "IAQ Index Accuracy: Low\n";
    }
    elsif ($iaq_index_accuracy == Tinkerforge::BrickletAirQuality->ACCURACY_MEDIUM)
    {
        print "IAQ Index Accuracy: Medium\n";
    }
    elsif ($iaq_index_accuracy == Tinkerforge::BrickletAirQuality->ACCURACY_HIGH)
    {
        print "IAQ Index Accuracy: High\n";
    }

    print "Temperature: " . $temperature/100.0 . " °C\n";
    print "Humidity: " . $humidity/100.0 . " %RH\n";
    print "Air Pressure: " . $air_pressure/100.0 . " hPa\n";
    print "\n";
}

my $ipcon = Tinkerforge::IPConnection->new(); # Create IP connection
my $aq = Tinkerforge::BrickletAirQuality->new(&UID, $ipcon); # Create device object

$ipcon->connect(&HOST, &PORT); # Connect to brickd
# Don't use device before ipcon is connected

# Register all values callback to subroutine cb_all_values
$aq->register_callback($aq->CALLBACK_ALL_VALUES, 'cb_all_values');

# Set period for all values callback to 1s (1000ms)
$aq->set_all_values_callback_configuration(1000, 0);

print "Press key to exit\n";
<STDIN>;
$ipcon->disconnect();
